src/Simple/JSON.purs

```diff
@@ -96,3 +103,33 @@ instance readFieldsNil ::
   ) => ReadForeignFields Nil row where
   getFields _ _ _ =
     pure $ to {}
+
+
+-- | The purpose of the `ReadKey` class is to turn strings into map keys,
+-- | so we can read generic `Map`s from `StrMap`s.
+class ReadKey a where
+  readKey :: String -> F a
```

Comment:
You might want to use a map whose keys are sum types. This class would allow you to convert the strings in the JSON into values of your sum type.